### PR TITLE
Typescript Bug Fixes

### PR DESCRIPTION
**Pull request by mrtracy**
The typescript definition file `mithril.d.ts` had several bugs which prevented several common Mithril use cases from compiling in typescript.  Three commits in this pull request:

* Repair the definition of MithrilRoutes, which previously would not compile unless all Routes to used the same controller type.
* Repair the definition of MithrilVirtualElement, which previously would not compile unless all Components defined in the tree used the same controller type.
* Add support for Parameterized components, which the Typescript definition previously would not compile.

### CODE REVIEW

File: mithril.d.ts [hunk 1]
```diff
@@ -59,26 +43,103 @@ declare module _mithril {
...
 		*
 		* @see m.render
 		* @see m.mount
 		* @see m.component
 		*/
-		<T extends MithrilController>(
-			component: MithrilComponent<T>,
-			...args: any[]
-		): MithrilComponent<T>;
+		<TController extends MithrilController>(
```
Comment (impinball) [inline mithril.d.ts]: This is okay to keep.

**Comment by impinball** [comment]
@mrtracy 

-1 for most of this, as it loosens the types significantly. The `<T extends MithrilController>` parametrization ensures that the controllers are actually instances of MithrilController.

If this is actually fixing anything (other than what I've explicitly noted is okay to keep), please link to a relevant example in the [TypeScript playground](http://www.typescriptlang.org/Playground) that demonstrates a type error caused by this.

File: mithril.d.ts [hunk 2]
```diff
@@ -149,18 +210,92 @@ declare module _mithril {
...
+		* @see m.mount
+		* @see m
+		*/
+		component<TController extends MithrilController, T1, T2, T3, T4>(
+			component: ParameterizedMithrilComponent<TController, T1, T2, T3, T4>,
+			arg1: T1,
+			arg2: T2,
+			arg3: T3,
+			arg4: T4
+		): MithrilComponent<TController>;
```
Comment (impinball) [inline mithril.d.ts]: Nit: keep the base `...args: any` type that was initially here.

This is probably better to separate into another PR.

File: mithril.d.ts [hunk 3]
```diff
@@ -596,7 +731,32 @@ declare module _mithril {
...
+	* @see m.component
+	*/
+	interface ParameterizedMithrilComponent<TController extends MithrilController, T1, T2, T3, T4> {
+		/**
+		* The component's controller.
+		*
+		* @see m.component
+		*/
+		controller: (arg1: T1, arg2: T2, arg3: T3, arg4: T4) => TController |
+		  {new(arg1: T1, arg2: T2, arg3: T3, arg4: T4): TController};
```
Comment (impinball) [inline mithril.d.ts]: Don't use `new` for this part. Use the function call form for now. Currently, the controller is created with `.apply()`, and the prototype is set to the component itself.

File: mithril.d.ts [hunk 4]
```diff
@@ -596,7 +731,32 @@ declare module _mithril {
...
+	}
+
+	/**
+	* This represents a parameterized MithrilComponent with up to six parameters,
+	* intended to be instantiated via m.component().
+	*
+	* @see m
+	* @see m.component
+	*/
+	interface ParameterizedMithrilComponent<TController extends MithrilController, T1, T2, T3, T4> {
```
Comment (impinball) [inline mithril.d.ts]: I'm okay with this, as long as there's a base case of `...args: any[]` to soak up anything else. Pull this out into its own separate PR.

**Comment by mrtracy** [comment]
I'm going to close this pull request and open two separate ones: one for the parameterized component addition, one for the proposed changes to MithrilRoutes and MithrilVirtualElement

**Comment by mrtracy** [comment]
#806 and #807
